metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowPostProcessor.java

```diff
@@ -50,49 +66,91 @@ public WorkflowPostProcessor(DepublishRecordIdDao depublishRecordIdDao,
-            .getById(latestSuccessfulPlugin.getExecutionId());
-        final Optional<AbstractMetisPlugin> metisPluginWithType = workflowExecutionToUpdate
-            .getMetisPluginWithType(latestSuccessfulPlugin.getPlugin().getPluginType());
-        if (metisPluginWithType.isPresent()) {
-          metisPluginWithType.get().setDataStatus(DataStatus.DELETED);
-          workflowExecutionDao.updateWorkflowPlugins(workflowExecutionToUpdate);
-        }
+  private void depublishPostProcess(DepublishPlugin plugin, String datasetId) throws DpsException {
+    if (plugin.getPluginMetadata().isDatasetDepublish()) {
+      depublishDatasetPostProcess(datasetId);
+    } else {
+      depublishRecordPostProcess(plugin, datasetId);
+    }
+  }
+
+  private void depublishDatasetPostProcess(String datasetId){
+
+    // Set all depublished records back to PENDING.
+    depublishRecordIdDao.markRecordIdsWithDepublicationStatus(datasetId, null,
+            DepublicationStatus.PENDING_DEPUBLICATION, null);
+
+    // Find latest PUBLISH Type Plugin and set dataStatus to DELETED.
+    final PluginWithExecutionId<MetisPlugin> latestSuccessfulPlugin = workflowExecutionDao
+            .getLatestSuccessfulPlugin(datasetId, OrchestratorService.PUBLISH_TYPES);
+    if (Objects.nonNull(latestSuccessfulPlugin) && Objects
+            .nonNull(latestSuccessfulPlugin.getPlugin())) {
+      final WorkflowExecution workflowExecutionToUpdate = workflowExecutionDao
+              .getById(latestSuccessfulPlugin.getExecutionId());
+      final Optional<AbstractMetisPlugin> metisPluginWithType = workflowExecutionToUpdate
+              .getMetisPluginWithType(latestSuccessfulPlugin.getPlugin().getPluginType());
+      if (metisPluginWithType.isPresent()) {
+        metisPluginWithType.get().setDataStatus(DataStatus.DELETED);
+        workflowExecutionDao.updateWorkflowPlugins(workflowExecutionToUpdate);
       }
-      final Dataset dataset = datasetDao.getDatasetByDatasetId(datasetId);
-      dataset.setPublicationFitness(PublicationFitness.UNFIT);
+    }
+
+    // Set publication fitness to UNFIT.
+    final Dataset dataset = datasetDao.getDatasetByDatasetId(datasetId);
+    dataset.setPublicationFitness(PublicationFitness.UNFIT);
+    datasetDao.update(dataset);
+  }
+
+  private void depublishRecordPostProcess(DepublishPlugin depublishPlugin, String datasetId)
+          throws DpsException {
+
+    // Retrieve the successfully depublished records.
+    final long externalTaskId = Long.parseLong(depublishPlugin.getExternalTaskId());
+    final List<SubTaskInfo> subTasks = new ArrayList<>();
+    while (true) {
+      final List<SubTaskInfo> subTasksBatch = dpsClient.getDetailedTaskReportBetweenChunks(
+              depublishPlugin.getTopologyName(), externalTaskId, subTasks.size(),
+              subTasks.size() + ECLOUD_REQUEST_BATCH_SIZE);
+      subTasks.addAll(subTasksBatch);
+      if (subTasksBatch.size() < ECLOUD_REQUEST_BATCH_SIZE) {
```

Comment:
Integrate this into the loop I would say

Reply:
Ok. Like so?

Reply:
Yeap!